sklearn/utils/estimator_checks.py

```diff
@@ -3421,3 +3421,99 @@ def check_estimator_get_tags_default_keys(name, estimator_orig):
         f"{name}._get_tags() is missing entries for the following default tags"
         f": {default_tags_keys - tags_keys.intersection(default_tags_keys)}"
     )
+
+
+def check_dataframe_column_names_consistency(name, estimator_orig):
+    try:
+        import pandas as pd
+    except ImportError:
+        raise SkipTest(
+            "pandas is not installed: not checking "
+            "column name consistency for pandas"
+        )
+
+    def _construct_dataframe(X, columns):
+        return pd.DataFrame(X, columns=columns)
+
+    _check_column_name_consistency(
+        name, estimator_orig, _construct_dataframe, "dataframe"
+    )
+
+
+def _check_column_name_consistency(name, estimator_orig, construct_X, array_name):
+    tags = _safe_tags(estimator_orig)
+
+    if (
+        "2darray" not in tags["X_types"]
+        and "sparse" not in tags["X_types"]
+        or tags["no_validation"]
+    ):
+        return
+
+    rng = np.random.RandomState(0)
+
+    estimator = clone(estimator_orig)
+    set_random_state(estimator)
+    if "warm_start" in estimator.get_params():
+        estimator.set_params(warm_start=False)
+
+    X_orig = rng.normal(size=(150, 8))
+    X_orig = _enforce_estimator_tags_x(estimator, X_orig)
+    X_orig = _pairwise_estimator_convert_X(X_orig, estimator)
+    n_samples, n_features = X_orig.shape
+
+    names = np.array([f"col_{i}" for i in range(n_features)])
+    X = construct_X(X_orig, names)
+
+    if is_regressor(estimator):
+        y = rng.normal(size=n_samples)
+    else:
+        y = rng.randint(low=0, high=2, size=n_samples)
+    y = _enforce_estimator_tags_y(estimator, y)
+
+    estimator.fit(X, y)
+
+    if not hasattr(estimator, "feature_names_in_"):
+        raise ValueError(
+            "Estimator does not have a feature_names_in_ "
+            f"attribute after fitting with a {array_name}"
+        )
+
+    check_funcs = [
+        getattr(estimator, func)
+        for func in ("predict", "transform", "decision_function", "predict_proba")
```

Comment:
add `inverse_transform` and `predict_log_proba`?

Reply:
`inverse_transform` would not work because estimators all output ndarrays. I added `predict_log_proba`.